src/MDOP.jl

```diff
@@ -4,16 +4,150 @@ using Distances
+statistic `FNNs` [^Hegger1999][^Kennel1992] is being checked and as soon as this statistic
+drops below the threshold `fnn_thres`, the algorithm breaks. In order to
+increase the  practability of the method the algorithm also breaks, when the
+FNN-statistic `FNNs` increases . The final embedding vector is stored in `Y`
+(`Dataset`). The chosen delay values for each embedding cycle are stored in the
+`τ_vals` and the according time series number chosen for the according delay
+value in `τ_vals` is stored in `ts_vals`. For univariate embedding (`s::Vector`)
+`ts_vals` is a vector of ones of length `τ_vals`, because there is simply just
+time series to choose from. If `βs=true` then the function also returns the
+`β`-statistic for each embedding cycle as a `List` of `Vector`s.
 
 [^Nichkawde2013]: Nichkawde, Chetan (2013). [Optimal state-space reconstruction using derivatives on projected manifold. Physical Review E 87, 022905](https://doi.org/10.1103/PhysRevE.87.022905).
+[^Hegger1999]: Hegger, Rainer and Kantz, Holger (1999). [Improved false nearest neighbor method to detect determinism in time series data. Physical Review E 60, 4970](https://doi.org/10.1103/PhysRevE.60.4970).
+[^Kennel1992]: Kennel, M. B., Brown, R., Abarbanel, H. D. I. (1992). [Determining embedding dimension for state-space reconstruction using a geometrical construction. Phys. Rev. A 45, 3403] (https://doi.org/10.1103/PhysRevA.45.3403).
 """
-function MDOP(Y...)
+function MDOP(s::Vector{Float64}; τs = 0:50 , w::Int = 1, fnn_thres::Float64 = 0.05,
+    r::Float64 = 2.0, βs::Bool=false)
+    @assert 0 < fnn_thres < 1 "Please select a valid breaking criterion, i.e. a threshold value `fnn_thres` ∈ (0 1)"
+    @assert all(x -> x ≥ 0, τs)
+
+    max_num_of_cycles = 50 # assure that the algorithm will break after 50 embedding cycles
+
+    metric = Euclidean()
+    # normalize input time series (especially important for fnn-computation)
+    s = (s.-mean(s))./std(s)
+    # define actual phase space trajectory
+    Y_act = Dataset(s)
+
+    # compute nearest neighbor distances in the first embedding dimension for
+    # FNN statistic
+    vtree = KDTree(Y_act, metric)
+    allNNidxs, NNdist_old = all_neighbors(vtree, Y_act, 1:length(Y_act), 1, w)
+
+    # set a flag, in order to tell the while loop when to stop. Each loop
+    # stands for encountering a new embedding dimension
+    flag = true
+
+    # set index-counter for the while loop
+    cnt = 1;
+
+    # preallocate output variables
+    τ_vals = zeros(Int,1)
+    ts_vals = ones(Int,1)
+    FNNs = zeros(Float64,1)
+    βs ? βS = Array{Float64, 2}(undef, length(τs), max_num_of_cycles) : nothing
+
+
+    # loop over increasing embedding dimensions until some break criterion will
+    # tell the loop to stop/break
+    while flag
+
+        # get the β-statistic
+        β = beta_statistic(Y_act, s; τs = τs , w = w)
+        βs ? βS[:,cnt] = β : nothing
```

Comment:
I don't understand what you are trying to do here: `βs ? βS[:,cnt] = β : nothing` ? It is untypical in Julia to do assignment in such a manner. (same with doing `[s[i] = 3 for i in 1:length(s)]`. It is unintuitive syntax, because above the `nothing` is not used anywhere, while in the second example the list comprehension `[ ]` is also not used, since you don't use the created array.

Do instead:
```julia
if βs
  βS[:, cnt] .= β
end
```

I also have another request: `βs` and `βS` have variable names that appear very similar in code, yet their nature is completely different: one is a boolean, the other is a matrix. How about naming the boolean one something like `β_bool` or `β_condition`, or in general something that makes sense with respect to the algorithm?

Reply:
accepted